File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/the_hurdle_race.c
// https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

#include <stdio.h>

void read_numbers(int *arr, size_t n);
int hurdle_race(const int *hurdles, size_t n, int max_jump);

int main()
{
    int n, max_jump;
    scanf("%d %d", &n, &max_jump);
    int hurdles[n];
    read_numbers(hurdles, n);

    printf("%d\n", hurdle_race(hurdles, n, max_jump));

    return 0;
}

void read_numbers(int *arr, size_t n)
{
    for (size_t i = 0; i < n; ++i)
        scanf("%d", &arr[i]);
}

// n: length of hurdles
// T: O(n)
// S: O(1) extra space
int hurdle_race(const int *hurdles, size_t n, int max_jump)
{
    int highest = hurdles[0];
    for (size_t i = 1; i < n; ++i)
        highest = highest > hurdles[i] ? highest : hurdles[i];

    int max = highest - max_jump;
    return 0 > max ? 0 : max;
}

File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/the_hurdle_race.go
// https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

package main

import (
    "fmt"
    "slices"
)

func main() {
    var n, maxJump int
    _, _ = fmt.Scan(&n, &maxJump)
    hurdles := readNumbers(n)
    fmt.Println(hurdleRace(hurdles, maxJump))
}

func readNumbers(n int) []int {
    arr := make([]int, n)
    for i := range arr {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// n: length of array hurdles
// T: O(n)
// S: O(1) extra space
func hurdleRace(hurdles []int, maxJump int) int {
    return max(0, slices.Max(hurdles) - maxJump)
}
